src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java

```diff
@@ -165,6 +172,131 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+    public static class BuildingBuilderResource
+    {
+        /**
+         * Availability status of the resource.
+         * according to the builder's chest, inventory and the player's inventory
+         */
+        public enum RessourceAvailability
+        {
+            NOT_NEEDED,
+            DONT_HAVE,
+            NEED_MORE,
+            HAVE_ENOUGHT
+        }
+
+        private int amountAvailable;
+        private int amountPlayer;
+        private final String name;
+        private final ItemStack itemStack;
+        private final int amountNeeded;
+
+        BuildingBuilderResource(final String name, final ItemStack itemStack, final int amountAvailable, final int amountNeeded)
+        {
+            this.name = name;
+            this.itemStack = itemStack;
+            this.amountAvailable = amountAvailable;
+            this.amountNeeded = amountNeeded;
+            this.amountPlayer = 0;
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public ItemStack getItemStack()
+        {
+            return itemStack;
+        }
+
+        /**
+         * get the amount available for this resource.
+         * i.e. amount in the chest + amount in the builder's inventory
+         */
+        public int getAvailable()
+        {
+            return amountAvailable;
+        }
+
+        public void setAvailable(final int amount)
+        {
+            amountAvailable=amount;
+        }
+
+        public int getNeeded()
+        {
+            return amountNeeded;
+        }
+
+        public void setPlayerAmount(final int amount)
+        {
+            amountPlayer = amount;
+        }
+
+        public int getPlayerAmount()
+        {
+            return amountPlayer;
+        }
+
+        public RessourceAvailability getAvailabilityStatus()
+        {
+            if (amountNeeded >amountAvailable)
+            {
+                if (amountPlayer==0)
+                {
+                    return RessourceAvailability.DONT_HAVE;
+                }
+                if (amountPlayer<(amountNeeded-amountAvailable))
+                {
+                    return RessourceAvailability.NEED_MORE;
+                }
+                return RessourceAvailability.HAVE_ENOUGHT;
+            }
+            return RessourceAvailability.NOT_NEEDED;
+        }
+
+        @Override
+        public String toString()
+        {
+            final int itemId=Item.getIdFromItem(itemStack.getItem());
+            final int damage=itemStack.getItemDamage();
+            return name + "(p:"+amountPlayer+" a:" +amountAvailable+" n:"+amountNeeded+" id="+itemId+" damage="+damage+") => "+getAvailabilityStatus().name();
+        }
+
+
+        /**
+         * Comparator class for BuildingBuilderResource.
+         *
+         * This is use in the gui to order the list of resources needed.
+         */
+        public static class ResourceComparator implements Comparator<BuildingBuilderResource>
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Make this class "Serializable". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2063)